rust/datafusion/src/execution/physical_plan/sort.rs

```diff
@@ -0,0 +1,211 @@
+                expr: self.expr.clone(),
+                schema: self.schema(),
+            })),
+        ])
+    }
+}
+
+/// Represents a single partition of a Sort execution plan
+struct SortPartition {
+    schema: Arc<Schema>,
+    expr: Vec<PhysicalSortExpr>,
+    input: Vec<Arc<dyn Partition>>,
+}
+
+impl Partition for SortPartition {
+    /// Execute the sort
+    fn execute(&self) -> Result<Arc<Mutex<dyn BatchIterator>>> {
+        let threads: Vec<JoinHandle<Result<Vec<RecordBatch>>>> = self
+            .input
+            .iter()
+            .map(|p| {
+                let p = p.clone();
+                thread::spawn(move || {
+                    let it = p.execute()?;
+                    common::collect(it)
+                })
+            })
+            .collect();
+
+        // generate record batches from input in parallel
+        let mut all_batches: Vec<Arc<RecordBatch>> = vec![];
+        for thread in threads {
+            let join = thread.join().expect("Failed to join thread");
+            let result = join?;
+            result
+                .iter()
+                .for_each(|batch| all_batches.push(Arc::new(batch.clone())));
+        }
+
+        // combine all record batches into one for each column
+        let combined_batch = RecordBatch::try_new(
+            self.schema.clone(),
+            self.schema
+                .fields()
+                .iter()
+                .enumerate()
+                .map(|(i, _)| -> Result<ArrayRef> {
+                    Ok(concat(
+                        &all_batches
+                            .iter()
+                            .map(|batch| batch.columns()[i].clone())
+                            .collect::<Vec<ArrayRef>>(),
+                    )?)
+                })
+                .collect::<Result<Vec<ArrayRef>>>()?,
+        )?;
+
+        // sort combined record batch
+        let indices = lexsort_to_indices(
+            &self
+                .expr
+                .iter()
+                .map(|e| e.evaluate_to_sort_column(&combined_batch))
+                .collect::<Result<Vec<SortColumn>>>()?,
+        )?;
+
+        // reorder all rows based on sorted indices
+        let sorted_batch = RecordBatch::try_new(
+            self.schema.clone(),
+            combined_batch
+                .columns()
+                .iter()
+                .map(|column| -> Result<ArrayRef> {
+                    Ok(take(
+                        column,
+                        &indices,
+                        // disable bound check overhead since indices are already generated from
+                        // the same record batch
+                        Some(TakeOptions {
+                            check_bounds: false,
+                        }),
+                    )?)
+                })
+                .collect::<Result<Vec<ArrayRef>>>()?,
+        )?;
+
+        Ok(Arc::new(Mutex::new(RecordBatchIterator::new(
```

Comment:
More a question for @andygrove; in the case of a sort operation consolidating all partitions into a single larger partition, would we need to repartition the record batches?

Reply:
I have the same question while developing this, curious what @andygrove thoughts on this as well. I left it as one single batch for performance reason (to avoid having to run the split code).